cpp/tests/shared_fcntl_mutex_test.cc

```diff
@@ -0,0 +1,516 @@
+  ASSERT_OK(comms[0].expect("ok", PARENT_CONTEXT));
+
+  // Second process observes no lock
+  ASSERT_OK(comms[1].send(kRequestLock + " read"s, PARENT_CONTEXT));
+  ASSERT_OK(comms[1].expect("none", PARENT_CONTEXT));
+
+  // Second process acquires a read lock
+  ASSERT_OK(comms[1].send(kReadLock + " thread 1"s, PARENT_CONTEXT));
+  ASSERT_OK(comms[1].expect("ok", PARENT_CONTEXT));
+
+  // First process observes no lock when requesting a read lock
+  ASSERT_OK(comms[0].send(kRequestLock + " read"s, PARENT_CONTEXT));
+  ASSERT_OK(comms[0].expect("none", PARENT_CONTEXT));
+
+  // First process observes read lock when requesting a write lock
+  ASSERT_OK(comms[0].send(kRequestLock + " write"s, PARENT_CONTEXT));
+  auto child_1_pid = std::to_string(child_pid[1]);
+  ASSERT_OK(comms[0].expect("fail read " + child_1_pid, PARENT_CONTEXT));
+
+  ASSERT_OK(ShutdownChildren(comms, child_pid));
+}
+
+// Test readonly fallback in the case of a
+// read-only table lock/directory
+TEST(SharedFcntlMutexTest, ReadFallback) {
+  using fs::perm_options;
+  using fs::perms;
+
+  auto tmp_dir = temp_directory_name();
+
+  // Cleanup test directory
+  std::shared_ptr<void> defer(nullptr, [&](...) {
+    std::filesystem::permissions(tmp_dir,
+                                 perms::owner_all | perms::group_all | perms::others_all,
+                                 perm_options::add);
+
+    fs::remove_all(tmp_dir);
+  });
+
+  ASSERT_TRUE(fs::create_directory(tmp_dir));
+  std::filesystem::permissions(tmp_dir,
+                               perms::owner_all | perms::group_all | perms::others_all,
+                               perm_options::remove);
+
+  auto lock_name = tmp_dir / "arcae.table.lock";
+
+  ASSERT_OK_AND_ASSIGN(auto lock, SharedFcntlMutex::Create(lock_name.native()));
+  ASSERT_FALSE(lock->has_fd());
+  // Write locks fail but read locks succeed
+  ASSERT_NOT_OK(lock->lock());
+  ASSERT_OK(lock->lock_shared());
+}
+
+// Test that the SharedFcntlGuard behaves as expected
+TEST(SharedFcntlMutexTest, LockGuard) {
+  auto path = temp_directory_name();
+  if (!fs::create_directory(path)) FAIL() << "Unable to create " << path.native();
+  auto lock_name = path / "table.lock";
+
+  ASSERT_OK_AND_ASSIGN(auto mutex, SharedFcntlMutex::Create(lock_name.native()));
+  bool success;
+
+  {
+    // Read guard
+    arcae::SharedFcntlGuard lock(*mutex, false);
+    // Write locking fails
+    ASSERT_OK_AND_ASSIGN(success, mutex->try_lock());
+    ASSERT_FALSE(success);
+
+    ASSERT_EQ(mutex->fcntl_readers(), 1);
+  }
+
+  // Can write lock
+  ASSERT_OK_AND_ASSIGN(success, mutex->try_lock());
+  ASSERT_TRUE(success);
+  mutex->unlock();
+
+  {
+    // Write guard
+    arcae::SharedFcntlGuard lock(*mutex, true);
+    // Read locking fails
+    ASSERT_OK_AND_ASSIGN(success, mutex->try_lock_shared());
+    ASSERT_FALSE(success);
+  }
+
+  // Can read lock lock
+  ASSERT_OK_AND_ASSIGN(success, mutex->try_lock_shared());
+  ASSERT_TRUE(success);
+  mutex->unlock();
+}
```

Comment:
You don't seem to have any tests for the contended case (in particular, that exponential backoff code). There are doesn't seem to be much if any testing that holding a read lock prevents write locks being taken in other threads or processes.